drape_frontend/animation_system.cpp

```diff
@@ -0,0 +1,1160 @@
+}
+
+void PerspectiveSwitchAnimation::Advance(double elapsedSeconds)
+{
+  m_angleInterpolator->Advance(elapsedSeconds);
+}
+
+void PerspectiveSwitchAnimation::Finish()
+{
+  m_angleInterpolator->Finish();
+  Animation::Finish();
+}
+
+void PerspectiveSwitchAnimation::OnStart()
+{
+  if (m_isEnablePerspectiveAnim)
+    m_needPerspectiveSwitch = true;
+  Animation::OnStart();
+}
+
+void PerspectiveSwitchAnimation::OnFinish()
+{
+  if (!m_isEnablePerspectiveAnim)
+    m_needPerspectiveSwitch = true;
+  Animation::OnFinish();
+}
+
+void PerspectiveSwitchAnimation::SetMaxDuration(double maxDuration)
+{
+  m_angleInterpolator->SetMaxDuration(maxDuration);
+}
+
+double PerspectiveSwitchAnimation::GetDuration() const
+{
+  return m_angleInterpolator->GetDuration();
+}
+
+bool PerspectiveSwitchAnimation::IsFinished() const
+{
+  return m_angleInterpolator->IsFinished();
+}
+
+bool PerspectiveSwitchAnimation::GetProperty(TObject object, TProperty property, PropertyValue & value) const
+{
+  ASSERT(object == Animation::MapPlane, ());
+
+  switch (property)
+  {
+  case Animation::AnglePerspective:
+    value = PropertyValue(m_angleInterpolator->GetAngle());
+    return true;
+  case Animation::SwitchPerspective:
+    if (m_needPerspectiveSwitch)
+    {
+      m_needPerspectiveSwitch = false;
+      value = PropertyValue(SwitchPerspectiveParams(m_isEnablePerspectiveAnim, m_startAngle, m_endAngle, m_angleFOV));
+      return true;
+    }
+    return false;
+  default:
+    ASSERT(!"Wrong property", ());
+  }
+
+  return false;
+}
+
+ParallelAnimation::ParallelAnimation()
+  : Animation(true /* couldBeInterrupted */, true /* couldBeBlended */)
+{}
+
+Animation::TObjectProperties const & ParallelAnimation::GetProperties(TObject object) const
+{
+  ASSERT(HasObject(object), ());
+  return m_properties.find(object)->second;
+}
+
+bool ParallelAnimation::HasProperty(TObject object, TProperty property) const
+{
+  if (!HasObject(object))
+    return false;
+  TObjectProperties properties = GetProperties(object);
+  return properties.find(property) != properties.end();
+}
+
+void ParallelAnimation::AddAnimation(drape_ptr<Animation> && animation)
+{
+  TAnimObjects const & objects = animation->GetObjects();
+  m_objects.insert(objects.begin(), objects.end());
+  for (auto const & object : objects)
+  {
+    TObjectProperties const & properties = animation->GetProperties(object);
+    m_properties[object].insert(properties.begin(), properties.end());
+  }
+  m_animations.push_back(move(animation));
+}
+
+void ParallelAnimation::OnStart()
+{
+  for (auto & anim : m_animations)
+    anim->OnStart();
+}
+
+void ParallelAnimation::OnFinish()
+{
+}
+
+void ParallelAnimation::Advance(double elapsedSeconds)
+{
+  auto iter = m_animations.begin();
+  while (iter != m_animations.end())
+  {
+    (*iter)->Advance(elapsedSeconds);
+    if ((*iter)->IsFinished())
+    {
+      (*iter)->OnFinish();
+      iter = m_animations.erase(iter);
```

Comment:
m_animations is a deque, and erase is not cheap. Why not to use std::list here or, at least, accumulate them at the end of the container and remove at the end of the loop at once?